lib/tar.ml

```diff
@@ -661,230 +661,157 @@ module Header = struct
-              with
-              | Error _ as e -> return e
-              | Ok () ->
-                really_write fd (Bytes.unsafe_to_string buffer) >>= fun () ->
-                let payload = header.Header.file_name ^ "\000" in
-                really_write fd payload >>= fun () ->
-                really_write fd (Header.zero_padding blank) >>= fun () ->
-                return (Ok ())
-            end else
-              return (Ok ())
-          end >>= function
-          | Error _ as e -> return e
-          | Ok () -> return (Ok ())
-      end else
-       return (Ok ())) >>= function
-    | Error _ as e -> return e
-    | Ok () ->
-      let buffer = Bytes.make Header.length '\000' in
-      match Header.marshal ~level buffer header with
-      | Error _  as e -> return e
-      | Ok () ->
-        really_write fd (Bytes.unsafe_to_string buffer) >>= fun () ->
-        return (Ok ())
-
-  let write_extended ?level ~link_indicator hdr fd =
-    let link_indicator_name = match link_indicator with
-      | Header.Link.PerFileExtendedHeader -> "paxheader"
-      | Header.Link.GlobalExtendedHeader -> "pax_global_header"
-      | _ -> assert false
+    Ok ({ t with state = `Real_header extended },
+        Some (`Skip (Header.compute_zero_padding_length x)),
+        None)
+  | `Real_header extended ->
+    let* x =
+      Result.map_error
+        (fun _ -> `Fatal `Corrupt_pax_header) (* NB better error *)
+        (Header.unmarshal ~extended data)
     in
-    let pax_payload = Header.Extended.marshal hdr in
-    let pax = Header.make ~link_indicator link_indicator_name
-                (Int64.of_int @@ String.length pax_payload) in
-    write_unextended ?level pax fd >>= function
-    | Error _ as e -> return e
-    | Ok () ->
-      really_write fd pax_payload >>= fun () ->
-      really_write fd (Header.zero_padding pax) >>= fun () ->
-      return (Ok ())
-
-  let write ?level header fd =
-    ( match header.Header.extended with
-      | None -> return (Ok ())
-      | Some e ->
-        write_extended ?level ~link_indicator:Header.Link.PerFileExtendedHeader e fd )
-    >>= function
-    | Error _ as e -> return e
-    | Ok () -> write_unextended ?level header fd
-
-  let write_global_extended_header global fd =
-    write_extended ~link_indicator:Header.Link.GlobalExtendedHeader global fd
-end
+    let t, hdr = construct_header t x in
+    Ok (t, Some (`Header hdr), None)
+  | `Next_longlink x ->
+    let name = String.sub data 0 (String.length data - 1) in
+    let next_longlink = if x.Header.link_indicator = Header.Link.LongLink then Some name else t.next_longlink in
+    let next_longname = if x.Header.link_indicator = Header.Link.LongName then Some name else t.next_longname in
+    Ok ({ t with next_longlink ; next_longname ; state = `Active false },
+        Some (`Skip (Header.compute_zero_padding_length x)),
+        None)
+  | `Active read_zero ->
+    match Header.unmarshal ?extended:t.global data with
```

Comment:
That mostly means that at the beginning, the user **must** provides a 512 bytes `string` at first (according to `Header‧unmarshal`), otherwise the user gets an error. Should we precise that the user must read a `tar` file per 512 bytes? Should we provide a streaming API?